src/rgw/driver/d4n/rgw_sal_d4n.cc

```diff
@@ -467,13 +468,26 @@ int D4NFilterBucket::list(const DoutPrefixProvider* dpp, ListParams& params, int
         }
         blocks[i].cacheObj.bucketName = this->get_bucket_id();
         ldpp_dout(dpp, 20) << "D4NFilterBucket::" << __func__ << " blocks[i].cacheObj.objName: " << blocks[i].cacheObj.objName << dendl;
+
+        if (return_blocks) {
+          dir_blocks_objNames.push_back(blocks[i].cacheObj.objName);
+        }
         j++;
       }
       auto ret = blockDir->get(dpp, blocks, y);
       if (ret < 0) {
         ldpp_dout(dpp, 0) << "D4NFilterBucket::" << __func__ << " blockDir->get() returned error: " << ret << dendl;
         return ret;
       }
+      if (return_blocks) {
+        for (unsigned int i = 0; i < dir_blocks_objNames.size(); ++i) {
```

Comment:
This loop is adjacent to the previous while loop, so the time complexity should not become exponential but please let me know if you see a more efficient way to construct `dir_blocks` without this extra loop